README.md
semester_sortierung Block
=============

This file is part of the block_semester_sortierung plugin for Moodle - <http://moodle.org/>

*Author:*    Simeon Naydenov, Katarzyna Potocka

*Copyright:* 2020 [Academic Moodle Cooperation](http://www.academic-moodle-cooperation.org)

*License:*   [GNU GPL v3 or later](http://www.gnu.org/copyleft/gpl.html)


Description
-----------

In the semester_sortierung block courses are grouped by their starting dates in a semester view on the MyMoodle
page.

The semester_sortierung block also displays all MyMoodle information regarding the course activities of the
individual courses.

The module will also feature a "My favourites" area, for which courses from different semesters can
be selected and they will then be listed at the top of the favourites area.

**WARNING: This plugin is deprecated.** The last maintained version of the Plugin will be for Moodle 3.9. After 3.9 we won’t maintain the plugin any further. 
The reason behind that is that now you can achieve similar functionality using the core Course Overview Block and its new filtering capabilities.  

For further information on how to use moodle core features check the course overview functionality and the filter options. 



Example
-------

Use the semester_sortierung module to display the winter and summer semesters with all your courses,
and the favourites area for quick access.


Requirements
------------

The plugin is available for Moodle 2.5+. This version is for Moodle 3.9.


Installation
------------

* Copy the module code directly to the blocks/semester_sortierung directory.

* Log into Moodle as administrator.

* Open the administration area (http://your-moodle-site/admin) to start the installation
  automatically.


Admin Settings
--------------

As an administrator you can set the default values instance-wide on the settings page for
administrators in the semester_sortierung block:

* Sort courses by semesters (checkbox) - controls whether courses are sorted at all
* Winter semester months (checkbox) - select which months should apply for the winter semester
* Show favourites (checkbox) - controls whether favourites on
* Archive (number) - group semesters that are older than the specified value 
* Autocollapse courses (number) - always show courses older than the specified number of months as collapse (only if there are performance issues) 
* Skip older events (number) - don't load events older than the specified number of months (only if there are performance issues)

Privacy API
-----------

The semester_sortierung block fully implements the Moodle Privacy API.

Web Service
-----------

The semester_sortierung block provides the following web service functions.

* Authentication - via token

  Token can be obtained in two ways:
  - manually from Moodle at `$CFG->wwwroot/admin/settings.php?section=webservicetokens`
  - by calling `$CFG->wwwroot/login/token.php?username=USERNAME&password=PASSWORD&service=SERVICENAME`
    (here `SERVICENAME` is semester_sortierung Web service)

* Requests

  Get courses - get all enrolled and active courses for a student
  - URL: `$CFG->wwwroot/webservice/rest/server.php?moodlewsrestformat=json&wsfunction=block_semester_sortierung_get_courses&wstoken={token}&userid={userid}`
  - Method: GET
  - Produces: application/json
  - Sample output: `[ {"coursename": "COURSE NAME", "semestercode": "2016S", "description": "COURSE DESCRIPTION", "coursenumber": "015709-2016S", "courseid": 1234567} ]`

  Get course details - get the course metadata
  - URL: `$CFG->wwwroot/webservice/rest/server.php?moodlewsrestformat=json&wsfunction=block_semester_sortierung_get_modules&wstoken={token}&userid={userid}&courseid={courseid}`
  - Method: GET
  - Produces: application/json
  - Sample output: `[ {"modname": "MODULE_NAME", "modinfo": "HTML_INFORMATION"} ]`

  Get modules - get all active modules
  - URL: `$CFG->wwwroot/webservice/rest/server.php?moodlewsrestformat=json&wsfunction=block_semester_sortierung_get_modules&wstoken={token}`
  - Method: GET
  - Produces: application/json
  - Image format: base64 png
  - Sample output: `[ {"modname": "MODULE NAME - UNIQUE ID", "title_en": "Module title in English", "title_de": "Module title in German", "image": "base64EncodedImage"} ]`



Bug Reports / Support
---------------------

We try our best to deliver bug-free plugins, but we can not test the plugin for every platform,
database, PHP and Moodle version. If you find any bug please report it on
[GitHub](https://github.com/academic-moodle-cooperation/moodle-blocks_semester_sortierung/issues).
Please provide a detailed bug description, including the plugin and Moodle version and, if
applicable, a screenshot.

You may also file a request for enhancement on GitHub. If we consider the request generally useful
and if it can be implemented with reasonable effort we might implement it in a future version.

You may also post general questions on the plugin on GitHub, but note that we do not have the
resources to provide detailed support.


License
-------

This plugin is free software: you can redistribute it and/or modify it under the terms of the GNU
General Public License as published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

The plugin is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License with Moodle. If not, see
<http://www.gnu.org/licenses/>.


Good luck and have fun!


CHANGELOG.txt
CHANGELOG
=========

3.9.1 (2020-08-05)
------------------
* version 3.9 is the last supported version of this plugin. Use the moodle core block  “course overview” and the filter options, in order to provide your users with a similar experience.


3.9.0 (2020-06-16)
------------------
* Moodle 3.9 compatible version


3.8.0 (2019-11-15)
------------------
* Moodle 3.8 compatible version

3.7.1 (2019-11-15)
------------------
* [FEATURE] - support filter in course fullname and summary  [github pull @lucaboesch]

3.7.0 (2019-05-20)
------------------
* Moodle 3.7 compatible version

3.6.1 (2019-05-07)
------------------
* [BUG] Fix github #10 - Fix tiny typo in one English language string. PULL request submitted from Germán Valero
* [BUG] Fix github #9 - Remove unnecessary space before semester title. PULL request submitted from Luca Bösch

3.6 (2018-12-10)
------------------
* Moodle 3.6 compatible version
* [BUG] #5865 (GitHub #8) - fix php Notice on fresh install
* [FEATURE] #5757 - implement additional Privacy API functions
* [FEATURE] #3844 - implement unread forum posts count

3.5.0 (2018-05-??)
------------------
* Moodle 3.5 compatible version
* [FEATURE] Implement Privacy API
* [FEATURE] remove german language strings

3.4.0 (2017-11-13)
------------------

* Moodle 3.4 compatible version
* [UPDATE] fix codechecker warnings and errors
* [FIXED] move user personal sorting settings to a separate db table due to user_preference length limitation
* [FEATURE] add "Archive" section
* [BUG] improve performance when loading course events by introducing two new settings - 'Autocollapse courses'
  and 'Skip older events'

3.3.0 (2017-05-14)
------------------

* Moodle 3.3 compatible version
* [FEATURE] use course events as in the core my_overview block
* [FEATURE] use Mustache templates to render course events
* [DEPRECATED] use of modname_print_overview()


3.2.0 (2017-02-05)
------------------

* Moodle 3.2 compatible version
* [FEATURE] #3782 Rename web service to semester_sortierung Web service
* [REMOVED] #3688 Deprecated $plugin->cron


3.1.1 (2016-09-21)
------------------

* [FIXED] #3694 Months in settings.php to work in all time zones


3.1.0 (2016-07-28)
------------------

* Moodle 3.1 compatible version
* [FEATURE] Add external web service
* [FIXED] #3574 Long course/assignment names
* Added behat tests


3.0.0 (2016-03-28)
------------------

* Moodle 3.0 compatible version
* [FIXED] #3108 German lang file


2.9 (2015-10-14)
----------------

* Moodle 2.9 compatible version
* [FIXED] #2713 Bug with adding course to favorites - course can be added more
  than once


2.8.1 (2015-09-18)
------------------

* [FIXED] #2647: Star icon layout


2.8 (2015-05-25)
----------------

* Moodle 2.8 compatible version
* [FEATURE] #1970 Add personal sorting capability
* [CHANGED] #2354 Fix code to pass code checker
* [FIXED] #2337 Enable user to delete the block in his own my moodle page


2.7 (2014-12-08)
----------------

* Moodle 2.7 compatible version
* [FIXED] #2072 Changed settings plugin name to block_semester_sortierung, fix
  bug with config
* [Fixed] version.php and file headers


2.6 (2014-08-27)
----------------

* First release for Moodle 2.6


externallib.php
<?php
// This file is part of block_semester_sortierung for Moodle - http://moodle.org/
//
// Moodle is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Moodle is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.

/**
 * External API functions
 *
 * @package       block_semester_sortierung
 * @author        Andreas Hruska (andreas.hruska@tuwien.ac.at)
 * @author        Katarzyna Potocka (katarzyna.potocka@tuwien.ac.at)
 * @author        Simeon Naydenov (moniNaydenov@gmail.com)
 * @copyright     2020 Academic Moodle Cooperation {@link http://www.academic-moodle-cooperation.org}
 * @license       http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */

defined('MOODLE_INTERNAL') || die;


require_once("$CFG->libdir/externallib.php");

class block_semester_sortierung_external extends external_api {

    /**
     * Describes the parameters for get_modules.
     *
     * @return external_function_parameters
     * @since Moodle 3.0
     */
    public static function get_modules_parameters() {
        return new external_function_parameters (
            array()
        );
    }

    /**
     * Returns a list of all active modules
     *
     * @return array the forum details
     * @since Moodle 3.0
     */
    public static function get_modules() {
        global $CFG, $DB, $OUTPUT;
        $modules = $DB->get_records('modules', array('visible' => 1));
        $result = array();
        $CFG->svgicons = false;
        foreach ($modules as $module) {
            $m = new stdClass;
            $m->name = $module->name;

            $m->title_en = (new lang_string('pluginname',
                                            'mod_' . $module->name,
                                            null,
                                            'en'))->out();
            $m->title_de = (new lang_string('pluginname',
                                            'mod_' . $module->name,
                                            null,
                                            'de'))->out();
            $m->image = base64_encode(file_get_contents($OUTPUT->image_url('icon', 'mod_' . $module->name)));
            if (empty($m->image)) {
                $imagepath = $CFG->dirroot . '/mod/' . $module->name . '/pix/icon.';
                if (file_exists($imagepath . 'png')) {
                    $m->image = base64_encode(file_get_contents($imagepath . 'png'));
                } else if (file_exists($imagepath . 'gif')) {
                    $m->image = base64_encode(file_get_contents($imagepath . 'gif'));
                }
            }
            $result[] = $m;
        }
        return $result;
    }

    /**
     * Describes the get_modules return value.
     *
     * @return external_single_structure
     * @since Moodle 2.5
     */
    public static function get_modules_returns() {
        return new external_multiple_structure(
            new external_single_structure(
                array(
                    'name' => new external_value(PARAM_TEXT, 'Short one-word name of module'),
                    'title_en' => new external_value(PARAM_TEXT, 'Module title in english'),
                    'title_de' => new external_value(PARAM_TEXT, 'Module title in german'),
                    'image' => new external_value(PARAM_RAW, 'Module image')
                ), 'module'
            )
        );
    }

    /**
     * Describes the parameters for get_courses.
     *
     * @return external_function_parameters
     * @since Moodle 3.0
     */
    public static function get_courses_parameters() {
        return new external_function_parameters (
            array('userid' => new external_value(PARAM_TEXT, 'user id', VALUE_REQUIRED))
        );
    }

    /**
     * Returns a list of user courses
     *
     * @return array the forum details
     * @since Moodle 3.0
     */
    public static function get_courses($userid) {
        global $CFG, $DB, $OUTPUT;
        require_once(__DIR__ . '/locallib.php');

        $user = $DB->get_record('user', array('username' => $userid));

        // Get the information about the enrolled courses.
        $courses = enrol_get_all_users_courses($user->id, false , 'id, fullname, shortname, idnumber, summary');

        $config = get_config('block_semester_sortierung');
        // Add the semester to each course.
        $courses = block_semester_sortierung_fill_course_semester($courses, $config);

        $result = array();
        foreach ($courses as $semester => $seminfo) {
            if ($semester == 'fav') {
                continue;
            }
            foreach ($seminfo['courses'] as $semcourse) {
                $course = new stdClass;
                $course->coursenumber = $semcourse->idnumber;
                $course->coursename = $semcourse->fullname;
                $course->description = $semcourse->summary;
                $course->courseid = $semcourse->id;
                $course->semestercode = $semcourse->semester_short;
                $result[] = $course;
            }
        }
        return $result;
    }

    /**
     * Describes the get_courses return value.
     *
     * @return external_single_structure
     * @since Moodle 2.5
     */
    public static function get_courses_returns() {
        return new external_multiple_structure(
            new external_single_structure(
                array(
                    'coursename' => new external_value(PARAM_TEXT, 'Name of course'),
                    'semestercode' => new external_value(PARAM_RAW, 'Semester code'),
                    'coursenumber' => new external_value(PARAM_RAW, 'Course number'),
                    'courseid' => new external_value(PARAM_INT, 'Database unique course id'),
                    'description' => new external_value(PARAM_RAW, 'Course description')
                ), 'course'
            )
        );
    }

    /**
     * Describes the parameters for get_coursedetails.
     *
     * @return external_function_parameters
     * @since Moodle 3.0
     */
    public static function get_coursedetails_parameters() {
        return new external_function_parameters (
            array(
                'userid' => new external_value(PARAM_TEXT, 'user id', VALUE_REQUIRED),
                'courseid' => new external_value(PARAM_INT, 'course id', VALUE_REQUIRED)
            )
        );
    }

    /**
     * Returns a list of user courses
     *
     * @return array the forum details
     * @since Moodle 3.0
     */
    public static function get_coursedetails($userid, $courseid) {
        global $CFG, $DB, $OUTPUT, $USER, $PAGE;
        require_once(__DIR__ . '/locallib.php');
        $user = $DB->get_record('user', array('username' => $userid));
        $courses = $DB->get_records('course', array('id' => $courseid));
        if (!$user || empty($courses)) {
            return array();
        }
        $context = \context_course::instance($courseid);
        self::validate_context($context);

        $olduser = $USER;
        $GLOBALS['USER'] = $user; // Note for codechecker - GLOBALS is a php reserved variable, can't really change it.
        $result = array();
        $output = $PAGE->get_renderer('block_semester_sortierung');
        $expandedevents = block_semester_sortierung_get_courses_events($courses, $output);
        if (isset($expandedevents[$courseid]) && !empty($expandedevents[$courseid]->events)) {

            foreach ($expandedevents[$courseid]->events as $event) {
                $mod = new \stdClass;
                $mod->modname = $event->modulename;
                $mod->modinfo = $output->render_from_template('block_semester_sortierung/course-event-list-item-clean', $event);
                $result[] = $mod;
            }
        }

        $GLOBALS['USER'] = $olduser; // Note for codechecker - GLOBALS is a php reserved variable, can't really change it.

        return $result;
    }


    /**
     * Describes the get_coursedetails return value.
     *
     * @return external_single_structure
     * @since Moodle 2.5
     */
    public static function get_coursedetails_returns() {
        return new external_multiple_structure(
            new external_single_structure(
                array(
                    'modname' => new external_value(PARAM_TEXT, 'Name of module'),
                    'modinfo' => new external_value(PARAM_RAW, 'Module information')
                ), 'module'
            )
        );
    }
}